planner/core/plan_cache_lru.go

```diff
@@ -0,0 +1,210 @@
+// Package core Copyright 2022 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+package core
+
+import (
+	"container/list"
+	"sync"
+
+	"github.com/pingcap/errors"
+	"github.com/pingcap/tidb/types"
+	"github.com/pingcap/tidb/util/hack"
+	"github.com/pingcap/tidb/util/kvcache"
+)
+
+// CacheEntry wraps Key and Value. It's the value of list.Element.
+type CacheEntry struct {
+	PlanKey   kvcache.Key
+	PlanValue kvcache.Value
+}
+
+// LRUPlanCache is a dedicated least recently used cache, JUST use for plan cache.
+type LRUPlanCache struct {
+	capacity uint
+	size     uint
+	// buckets replace the map in general LRU
+	buckets map[hack.MutableString]map[*list.Element]struct{}
+	cache   *list.List
+	// lock make cache thread safe
+	lock sync.Mutex
+
+	// pickFromBucket get one element from bucket according to the incoming function.The LRUPlanCache can not work if this is nil
+	pickFromBucket func(map[*list.Element]struct{}, []*types.FieldType) (*list.Element, bool)
+	// onEvict will be called if any eviction happened
+	onEvict func(kvcache.Key, kvcache.Value)
+}
+
+// NewLRUPlanCache creates a PCLRUCache object, whose capacity is "capacity".
+// NOTE: "capacity" should be a positive value.
+func NewLRUPlanCache(capacity uint, pickFromBucket func(map[*list.Element]struct{}, []*types.FieldType) (*list.Element, bool),
+	onEvict func(kvcache.Key, kvcache.Value)) *LRUPlanCache {
+	if capacity < 1 {
+		panic("capacity of LRU Cache should be at least 1.")
```

Comment:
For safety, better to return an error here instead of panic.
Panic may cause the TiDB server process to exit, which is too dangerous.